Updates the Getting Started Guide (#78)
NUbots/NUbook: src/book/03-guides/01-main-codebase/01-getting-started.mdx

Docker is a container system that allows you to install development tools and run builds in an isolated environment that won't affect the rest of your system. We use it to set up the NUbots build system and run builds and binaries.

You will need to perform the steps correctly to set up Docker for your operating system to complete the example.

<details>

> <summary>Ubuntu</summary>

Reviewer: Since we specify `focal`, should we change this heading to `Ubuntu 20.04`?

Also, this section has headings for Mint, so should that be included in the heading too? We are assuming that these people won't know to even click on this tab, because they won't know the relationship between mint and ubuntu.

Author: I've added info on what to change this too if you aren't using focal fossa. Also changed the drop-down heading to Ubuntu/Linux Mint.

1. Add Docker's stable repository and update the apt package index.

   ```bash
   curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
   sudo add-apt-repository \
     "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
     focal \
     stable"
   sudo add-apt-repository universe

>    sudo apt update

Reviewer: I think these two update commands should be added before the install commands (also, couldn't the install commands be merged into one command?), because I think mint users will run these commands, then it won't work, then they'll run the mint command, then they'll run the install commands, which won't work because these commands were ran before the special mint command.
I can see myself falling victim to this, because usually I am an idiot who doesn't read the next steps before running the current ones lmao

Author: Do you mean after? I've split up the commands into different steps, is it better?
On the one command thing, is that with `sudo add-apt-repository universe && sudo apt update` do you mean or something else? I'd rather not concatenate commands incase errors come up between commands.

Reviewer: I think where you have put step 2:
```sh
sudo add-apt-repository universe
sudo apt update
``` 
is perfect now.

The commands I was thinking could be merged would be 
```sh
sudo apt-get install docker-ce docker-ce-cli containerd.io
sudo apt-get install python3 python3-pip
```
becomes:
```sh
# Pretty sure this would work?
sudo apt-get install docker-ce docker-ce-cli containerd.io python3 python3-pip
```

Sidenote: that step says "Install the lastest version of Docker", then it proceeds to install docker, docker's dependencies (fair), and python too. Could perhaps just change it to saying something which encompasses both?

Reviewer: Looks like they have changed the steps to install Docker, so we'll need to add a few commands. I tried to test the combined command in a fresh ubuntu container, and now the steps we have listed won't work because we haven't added the GPG key for the repo...

https://docs.docker.com/engine/install/ubuntu/

Gonna have to add a couple of steps.

Something that's really nice though, is that they have added a command to the part where you change between focal and bionic so that we don't have to hardcode the version at all any more. Might not need divergent mint and ubuntu steps now.

Reviewer: On second thought, I reckon we could just change the guide to say "Go to Docker(link) and follow the steps to install it." at this stage (maybe with a comment saying that mint users should just follow the ubuntu steps).
That would make that `apt-get install <packages>` just `sudo apt-get install python3 python3-pip`

Author: If it's `$(lsb_release -cs)`, I think that's always been a thing, it just doesn't work on Linux Mint.
Are the extra steps just 1 and 2 of 'SET UP THE REPOSITORY'?
We could add a link, kind of like an 'in case this doesn't work check if the instructions have changed' kind of thing. But there's a lot of extra stuff on there, and multiple ways to do things... so probs good to still have instructions here.

Reviewer: The extra steps involve installing 5 more packages, then curling and adding the docker GPG key, checking the fingerprint to make sure you did those properly, then the steps we already have. I am not sure how long that fingerprint would be the same, so I am sceptical that we might have to update that regularly

Author: Yeah maybe we should just link to it then. Then specify what section to do. And add any notes, like how to find what Ubuntu distro your Linux Mint version is.

</Alert>

<Alert type='info'>

GitKraken will not allow you to open the repository outside of WSL. Use the steps described in this article to [set up your system to use GitKraken in WSL](https://medium.com/@chuckdries/installing-gitkraken-in-wsl-2-15bf6459f823).

</Alert>

<Alert type='info'>

To be able to copy and paste into the Ubuntu terminal, right-click on the bar at the top of the Ubuntu window. Then click on **Properties**, and enable **Use Ctrl+Shift+C/V as Copy/Paste** in the **Options** tab. Note that this is <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>C</kbd>/<kbd>V</kbd> not <kbd>Ctrl</kbd>+<kbd>C</kbd>/<kbd>V</kbd>.

</Alert>

</details>

### A Text Editor

This section is not required for our example, but it is a good step to do in case you need to edit code in the future.

You'll need a text editor to make changes to the code. We recommend [Visual Studio Code](https://code.visualstudio.com/Download), but feel free to use your favourite editor.

> To ensure a consistent [code style](https://en.wikipedia.org/wiki/Programming_style), NUbots uses `clang-format`, a tool that reformats all code files in the repository to have the same style, even if they're written by different people. For convenience, you can install the `clang-format` [extension](https://marketplace.visualstudio.com/items?itemName=xaver.clang-format) in Visual Studio Code to automatically format the files when you make changes.

Reviewer: While we're here... (lmk if you think this should be in a separate PR or if you disagree with it)
```suggestion
To ensure a consistent [code style](https://en.wikipedia.org/wiki/Programming_style), NUbots uses [clang-format](https://clang.llvm.org/docs/ClangFormat.html), a tool that reformats code files in the repository to have the same style, even if they're written by different people. For convenience, you can install the `clang-format` [extension](https://marketplace.visualstudio.com/items?itemName=xaver.clang-format) in Visual Studio Code to automatically format the files when you make changes.
```

Reviewer 2: Good ol Kip the [rider](https://en.wikipedia.org/wiki/Rider_(legislation)) 😅.

2. With the target platform selected, choose the binaries (called "roles") that you want to build. You do this by running the following command:

   ```bash
   ./b configure
   ./b configure -i
   ```

   Note that you will need to do `configure` without `-i` the very first time you do this. Using `-i` will show a menu that will allow you to choose the roles to build. Enable the roles you want to build and disable the ones you don't by highlighting them using using the <kbd>↑</kbd> and <kbd>↓</kbd> arrow keys and pressing the spacebar to toggle on and off.

>    If you are following the keyboardwalkfake example, you will need to make sure `keyboardwalkfake` is set to `true`. Find the list of items that are prefixed with `ROLE_` and set all to `false` except for `keyboardwalkfake`, which should be `true`. Do not change any other lines unless you know what you are doing.

Reviewer: ```suggestion
   If you are following the keyboardwalkfake example, you will need to make sure `keyboardwalkfake` is set to `true`. Find the list of items that are prefixed with `ROLE_` and set all to `false` except for `keyboardwalkfake`, which should be `true`. Do not change other lines unless you want to build a different role.
```

Author: This was more about the build options than the other roles

Reviewer: ahh true. I was just thinking that this might spook people a bit much, without telling them what they need to be spooked about. This means that they might know everything they need to know, but they wouldn't know that they did, so they might still be spooked (I am having a hard time articulating this in a way that makes sense...).

Kind of like how we frame `./b shell`, as something you [should basically never do](https://nubook.nubots.net/system/foundations/build-system#shell), even though it's quite handy a lot of the time. I have no idea why it's such a big deal, and since the documentation about it doesn't say why we should be careful, I still don't know if I am missing something. Although, I have used it a bunch, and it's really handy.

Author: Haha yeah sure. Maybe just something like 'You do not need to change any other build options'.

>    This will start the binary to run the stand script, but the robot won't move until you press the green middle button on its back.

Reviewer: Should we mention what the red button does too?

Author: Probably worth mentioning for safety reasons.
